File: testdata/fixtures/bench/javascript/vulnerable/a08_deserialization.ts
// Source: CVE-2017-5941 - node-serialize RCE
// Expected: GTSS-GEN-002 (Unsafe Deserialization)
// OWASP: A08:2021 - Software and Data Integrity Failures

import { Request, Response } from 'express';
import serialize from 'node-serialize';

export function loadSession(req: Request, res: Response): void {
  const sessionCookie = req.cookies.session;
  if (!sessionCookie) {
    res.status(401).json({ error: 'No session' });
    return;
  }
  const decoded = Buffer.from(sessionCookie, 'base64').toString();
  const sessionData = serialize.unserialize(decoded);
  res.json({ user: sessionData.username, role: sessionData.role });
}

export function importConfig(req: Request, res: Response): void {
  const configStr = req.body.config;
  const config = JSON.parse(configStr);
  res.json({ status: 'Config loaded', keys: Object.keys(config) });
}

File: testdata/fixtures/bench/ruby/vulnerable/a08_deserialization.rb
# Source: CWE-502 - Unsafe deserialization in Ruby via YAML/Marshal
# Expected: GTSS-GEN-002 (Unsafe Deserialization - YAML.load/Marshal.load)
# OWASP: A08:2021 - Software and Data Integrity Failures

require 'yaml'
require 'base64'

class SessionController < ApplicationController
  def restore
    session_data = params[:session]
    decoded = Base64.decode64(session_data)
    data = YAML.load(decoded)
    session[:user] = data[:user]
    session[:role] = data[:role]
    redirect_to dashboard_path
  end

  def import_config
    config_yaml = request.body.read
    config = YAML.load(config_yaml)
    Setting.update_all_from(config)
    render json: { status: 'imported', keys: config.keys }
  end

  def load_cache
    key = params[:key]
    raw = Rails.cache.read(key)
    obj = Marshal.load(raw) if raw
    render json: obj || { error: 'not found' }
  end
end
